### ISSUE TITLE
Ignore unreachable code for type alias assignments

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#3555]
**Is your feature request related to a problem? Please describe.**

I want to be able to make use of Pyright's excellent support for recursive types however as mypy doesn't support this yet I need to be able to disable it when running mypy.

The mypy recommended method for conditionally assigning a type alias AFAIK is to use an `if MYPY` condition, for example:

```py
MYPY: Literal[False] = False
if MYPY:
    Data = Union[str, "Mapping[str, Any]"]
else:
    Data = Union[str, "Mapping[str, Data]"]
```

This works in mypy as expected, however Pyright complains with this error:

```
error: Illegal type annotation: variable not allowed unless it is a type alias (reportGeneralTypeIssues)
```

**Describe the solution you'd like**

With the given code sample, Pyright can infer that the assignment to `Data` on line 3 will never run and should ignore it and treat the line 5 assignment as the only assignment to the type alias.

**Additional context**

Unfortunately, running both mypy and Pyright over the codebase is a hard requirement as I don't want users of the library to encounter spurious errors with mypy that they cannot fix themselves. I apologise if there's prior discussion surrounding this, I couldn't find any.

**Comment by rattrayalex** [comment]
The `--always-true`/`--always-false` config option sounds interesting.

I wonder if `typing` should offer a facility akin to `sys.platform` for typecheckers, given that types in Python rely on third party systems, eg;

```py
import typing

if typing.implementation == 'mypy':
  Data = Union[str, "Mapping[str, Any]"]
elif typing.implementation == 'pyright':
  Data = Union[str, "Mapping[str, Data]"]
else:
  Data = Any
```

**Comment by erictraut** [comment]
I've added a new configuration option called "defineConstant". You can use it to specify one or more identifier names and associate then with boolean or string values. If pyright encounters one of these defined constants within a conditional statement (e.g. `if not DEBUG:` or `if FOO_VARIANT == 'premium':`, it will use the constant value for reachability analysis for the code block(s) guarded by the conditional statement.

To use, add the following to your pyrightconfig.json file:
```
{
    "defineConstant": { "MYPY": false }
}
```

**Comment by RobertCraigie** [comment]
Thank you! Looks like `defineConstant` will be a very useful solution.

The only potential issue with this solution is for library code as users will also have to add the `defineConstant` setting to their pyright config to resolve types correctly. As the majority of Pyright users are through the LSP then I can imagine that most users won't be aware of this option or might not even know that Pyright is being used under the hood and users that don't enable type checking will just be confused to why they aren't getting autocomplete or enhanced editor features in certain cases.

I appreciate that the solution might be to just document our config but in an ideal world it would just work for users as well. Maybe a good solution for this case would be bundling a pyright config with the package and support using `defineConstant` for that library code, although I imagine that would introduce a lot of complexity?

**Comment by erictraut** [comment]
Yes, this solution is not appropriate for library code. The solution that @hmc-cs-mdrissi suggested would be more appropriate for library code. That will require going through a standards process. If you would like to propose a new feature in typing-sig and draft a new PEP, you're welcome to do so. It would require all of the major Python type checkers to implement the solution, so it's not something that will happen quickly.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.254, which I just published. It will also be included in the next release of pylance.
